File: agents/backend/onyx/server/features/character_clothing_changer_ai_openrouter_truthgpt/utils/performance.py
"""
Performance Utilities
=====================

Utilities for performance monitoring and optimization.
"""

import time
import functools
from typing import Callable, Dict, Any, Optional
from contextlib import contextmanager
from dataclasses import dataclass, field
from collections import defaultdict
import logging

logger = logging.getLogger(__name__)


@dataclass
class PerformanceMetric:
    """Performance metric data"""
    name: str
    duration: float
    timestamp: float
    metadata: Dict[str, Any] = field(default_factory=dict)


class PerformanceMonitor:
    """
    Performance monitor for tracking function execution times.
    
    Features:
    - Track execution times
    - Aggregate statistics
    - Performance reports
    """
    
    def __init__(self):
        """Initialize performance monitor"""
        self.metrics: list[PerformanceMetric] = []
        self.max_metrics = 1000
    
    def record(self, name: str, duration: float, metadata: Dict[str, Any] = None):
        """
        Record a performance metric.
        
        Args:
            name: Metric name
            duration: Duration in seconds
            metadata: Optional metadata
        """
        metric = PerformanceMetric(
            name=name,
            duration=duration,
            timestamp=time.time(),
            metadata=metadata or {}
        )
        
        self.metrics.append(metric)
        
        # Keep only last max_metrics
        if len(self.metrics) > self.max_metrics:
            self.metrics = self.metrics[-self.max_metrics:]
    
    def get_stats(self, name: Optional[str] = None) -> Dict[str, Any]:
        """
        Get performance statistics.
        
        Args:
            name: Optional metric name to filter by
            
        Returns:
            Dictionary with statistics
        """
        metrics = self.metrics
        if name:
            metrics = [m for m in metrics if m.name == name]
        
        if not metrics:
            return {
                "count": 0,
                "name": name or "all"
            }
        
        durations = [m.duration for m in metrics]
        
        return {
            "name": name or "all",
            "count": len(metrics),
            "min": min(durations),
            "max": max(durations),
            "avg": sum(durations) / len(durations),
            "total": sum(durations)
        }
    
    def clear(self):
        """Clear all metrics"""
        self.metrics.clear()


# Global performance monitor
_performance_monitor: Optional[PerformanceMonitor] = None


def get_performance_monitor() -> PerformanceMonitor:
    """Get or create performance monitor instance"""
    global _performance_monitor
    if _performance_monitor is None:
        _performance_monitor = PerformanceMonitor()
    return _performance_monitor


@contextmanager
def measure_performance(name: str, metadata: Dict[str, Any] = None):
    """
    Context manager for measuring performance.
    
    Usage:
        with measure_performance("my_operation"):
            # code to measure
            pass
    """
    start_time = time.time()
    try:
        yield
    finally:
        duration = time.time() - start_time
        monitor = get_performance_monitor()
        monitor.record(name, duration, metadata)


def track_performance(name: Optional[str] = None) -> Callable:
    """
    Decorator to track function performance.
    
    Usage:
        @track_performance("my_function")
        async def my_function():
            ...
    """
    def decorator(func: Callable) -> Callable:
        metric_name = name or func.__name__
        
        @functools.wraps(func)
        async def async_wrapper(*args, **kwargs):
            with measure_performance(metric_name):
                return await func(*args, **kwargs)
        
        @functools.wraps(func)
        def sync_wrapper(*args, **kwargs):
            with measure_performance(metric_name):
                return func(*args, **kwargs)
        
        import asyncio
        if asyncio.iscoroutinefunction(func):
            return async_wrapper
        return sync_wrapper
    
    return decorator

File: Frontier-Model-run-polyglot/scripts/truthgpt-model-builder/lib/performance.ts
/**
 * Performance monitoring utilities
 */

export interface PerformanceMetric {
  name: string
  startTime: number
  endTime?: number
  duration?: number
  metadata?: Record<string, any>
}

class PerformanceMonitor {
  private metrics: Map<string, PerformanceMetric> = new Map()
  private history: PerformanceMetric[] = []
  private maxHistory = 1000

  start(name: string, metadata?: Record<string, any>): string {
    const id = `${name}-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`
    this.metrics.set(id, {
      name,
      startTime: performance.now(),
      metadata,
    })
    return id
  }

  end(id: string, metadata?: Record<string, any>): PerformanceMetric | null {
    const metric = this.metrics.get(id)
    if (!metric) return null

    metric.endTime = performance.now()
    metric.duration = metric.endTime - metric.startTime
    if (metadata) {
      metric.metadata = { ...metric.metadata, ...metadata }
    }

    this.metrics.delete(id)
    this.history.push(metric)

    // Keep only last maxHistory
    if (this.history.length > this.maxHistory) {
      this.history.shift()
    }

    return metric
  }

  getMetric(id: string): PerformanceMetric | undefined {
    return this.metrics.get(id)
  }

  getHistory(name?: string, limit?: number): PerformanceMetric[] {
    let filtered = this.history
    if (name) {
      filtered = filtered.filter(m => m.name === name)
    }
    if (limit) {
      filtered = filtered.slice(-limit)
    }
    return filtered
  }

  getAverageDuration(name: string): number {
    const metrics = this.getHistory(name)
    if (metrics.length === 0) return 0

    const total = metrics.reduce((sum, m) => sum + (m.duration || 0), 0)
    return total / metrics.length
  }

  clear(): void {
    this.metrics.clear()
    this.history = []
  }

  export(): any {
    return {
      active: Array.from(this.metrics.values()),
      history: this.history,
      averages: this.getAverageDurations(),
    }
  }

  private getAverageDurations(): Record<string, number> {
    const names = new Set(this.history.map(m => m.name))
    const averages: Record<string, number> = {}
    for (const name of names) {
      averages[name] = this.getAverageDuration(name)
    }
    return averages
  }
}

export const performanceMonitor = new PerformanceMonitor()


